pytng/pytng.pyx

```diff
@@ -1,147 +1,1481 @@
+                                                    TNG_USE_HASH,
+                                                    & data,
+                                                    & n_particles,
+                                                    & stride_length,
+                                                    n_values_per_frame,
+                                                    & datatype)
+
+        else:
+            if debug:
+                printf("reading NON particle data \n")
+            n_atoms[0] = 1  # still used for some allocs
+            # read non particle data off disk with hash checking
+            stat = tng_gen_data_vector_interval_get(self._traj,
+                                                    block_id,
+                                                    TNG_FALSE,
+                                                    self._frame,
+                                                    self._frame,
+                                                    TNG_USE_HASH,
+                                                    & data,
+                                                    NULL,
+                                                    & stride_length,
+                                                    n_values_per_frame,
+                                                    & datatype)
+
+        if stat != TNG_SUCCESS:
+            printf(
+                "WARNING: critical failure in tng_gen_data_vector_get \n")
+            return TNG_CRITICAL
+
+        stat = tng_data_block_num_values_per_frame_get(
+            self._traj, block_id, n_values_per_frame)
+        if stat == TNG_CRITICAL:
+            printf(
+                """WARNING: critical failure in
+                tng_data_block_num_values_per_frame_get \n""")
+            return TNG_CRITICAL
+
+        # renew the MemoryWrapper instance to have the right size
+        # to hold the data that has come off disk and will be cast to double*
+        self._wrapper.renew(
+            sizeof(double) * n_values_per_frame[0] * n_atoms[0])
+
+        _wrapped_values = <double*> self._wrapper.ptr
+
+        if self.debug:
+            printf("realloc values array to %ld doubles and %ld bits long \n",
+                   n_values_per_frame[0] * n_atoms[0], n_values_per_frame[0]
+                   * n_atoms[0]*sizeof(double))
+
+        # convert the data that was read off disk into an array of doubles
+        self.convert_to_double_arr(
+            data, _wrapped_values, n_atoms[0], n_values_per_frame[0],
+            datatype, debug)
+
+        # get the compression of the current frame
+        stat = tng_util_frame_current_compression_get(self._traj,
+                                                      block_id,
+                                                      & codec_id,
+                                                      & local_prec)
+        if stat == TNG_CRITICAL:
+            printf(
```

Comment:
We tend top avoid print (or printf) and rather use warning and logger. Is there a downside to using `warnings.warn` inside cython?

Reply:
This function is declared `nogil` for performance reasons, meaning the python GIL is released. As a consequence no python objects or methods can be called. Open to any other solutions or just removing.

Reply:
addendum: you can use `raise`, `assert` and `print` but this requires Cython >= 3.0

Reply:
Ok with `printf` – calling code should deal with `TNG_CRITICAL` anyway.

Perhaps once the whole library stabilizes we comment out the printfs but for right now it will help. Maybe be consistent with the printf and always prefix them with "pytng: " or something like that?

Reply:
Sounds good, as you said propagating errors up to the python level should be the end solution. :)